Refactored all code

nirolevy opened the pull request:
Closes #9 
Closes #3

nirolevy committed:
Merge pull request #1 from mh-guinness-gmail-com/master

re-structure + npm support + argsparse (#2)

PolaDinitz commented on Pipfile:
```diff
@@ -5,10 +5,13 @@ verify_ssl = true
 
 [dev-packages]
 pylint = "*"
+rope = "*"
```
If you know of a specific matching version it's best to download that version... less bugs in the future.

nirolevy replied:
Not relevant for `dev-packages` atm (it's packages that vscode uses), I guessed for `packages`... Not sure if it's correct

nirolevy replied:
@PolaDinitz ?

PolaDinitz commented on .gitignore:
```diff
@@ -145,3 +145,7 @@ dmypy.json
+
+tmp/
+in/
+out/
```
newline

syncush commented on __main__.py:
```diff
@@ -1,64 +1,44 @@
 from src.packager import package
+from src.command_line import args
+from src.providers import providers
+from src.loggers.ConsoleLogger import ConsoleLogger
```
Should we consider using `logging` pypi package?

nirolevy replied:
We should. In this PR or as an issue?

syncush replied:
I think we should cover this in this PR since it is a simple addition

syncush replied:
```
from sys import platform

if platform == "linux" or platform == "linux2":
    # linux

elif platform == "darwin":
    # OS X

elif platform == "win32":
    # Windows...
```

nirolevy replied:
I didn't understand your comment but moved to logging

nirolevy replied:
@syncush ?

syncush replied:
I dropped a code snippet for detecting on what os are you running for deciding the value of the default temp folder

syncush commented on src/command_line.py:
```diff
@@ -0,0 +1,25 @@
+
+__parser.add_argument('--concurrency-per-core', '-c', type=int, default=2, dest='concurrency',
+                      help='Number of workers = concurrency * CPU_CORES_COUNT')
+__parser.add_argument('--temp-dir', '-t', type=str, default='./tmp',
+                      help='Base directory to place all downloaded files before bundling')
```
we should detect the os platform and have default values `%TEMP%` or `/tmp` set respectively

nirolevy replied:
I think we should move to mem based FS

syncush replied:
`pytorch` for instance weights 800Mb, we can't store it on RAM

nirolevy replied:
How often is it updated? How many heavy packages are there?
I can argue that if `pytorch` is so heavy, maybe we shouldn't add it to this tool for now, at least until we know costs of azure resources

nirolevy replied:
Because it is downloaded, bundled and then uploaded:
1.  It may be necessary to keep the packages in memory for some parts of this process even if we do use a temp fs
2.  It will be A LOT faster to use mem fs

syncush replied:
I agree that we might have a problem with Azure resources. In the long term, our code should handle both small and large packages (think about docker images).

about `mem fs` vs `HDD` when we do implement such a feature, we should also introduce a flag for choosing between the two.

nirolevy replied:
After pola will introduce a DB, this will be possible. In the meantime, can you agree that implementing tmpfs and folder permissions is premature?

syncush replied:
I agree

nirolevy replied:
Opened as #16

nirolevy replied:
@syncush ?

syncush replied:
resolved

PolaDinitz commented on __main__.py:
```diff
@@ -1,64 +1,44 @@
-parser.add_argument('--proxy', type=str2bool, default=False,
-                    help='Turn on if you use a proxy')
-args = parser.parse_args()
+providers = {provider['name']: provider['class'] for provider in providers}
```
space { ... }

syncush replied:
@nirnir93  correct me if I am wrong but according to autopep8 putting a space is forbidden

nirolevy replied:
https://www.python.org/dev/peps/pep-0008/#id27

PolaDinitz replied:
I hate python!

syncush commented on src/downloader.py:
```diff
@@ -1,15 +1,33 @@
+    def __get_output_file_path(self, provider: Provider, product: Product) -> str:
+        package_dir = os.path.join(
+            self.__base_dir, provider.name, product.name)
+        os.makedirs(package_dir, exist_ok=True)
```
what about `mode=0666`, default value is `0777` which is `read write execute`

nirolevy replied:
What's wrong with 0777?
(I will fix according the the result of [this](https://github.com/mh-guinness-gmail-com/vsix-downloader/pull/15#discussion_r404696027))

syncush replied:
As mentioned above, for now, it is ok, resolved.

PolaDinitz commented on __main__.py:
```diff
@@ -1,64 +1,44 @@
+    resolved_products = []
+    logger.log('Started resolving products')
+    for provider_name, value in vars(args).items():
+        if provider_name in providers and value == True:
```
try `if ... value:` or `if ... value is True`. I prefer the first option, just don't compare it to true.

syncush replied:
`value is true` is the equiv of `some_variable === true` in javascript `if value` might cause undesirable behavior

PolaDinitz replied:
https://docs.quantifiedcode.com/python-anti-patterns/readability/comparison_to_true.html

nirolevy replied:
Fixed

nirolevy replied:
@PolaDinitz ?

PolaDinitz commented on __main__.py:
```diff
@@ -1,64 +1,44 @@
+            provider = providers[provider_name](logger)
+            product_names = get_lines('{0}/{1}.list'.format(args.input_dir, provider.name))
+            resolved_products += provider.provide(
+                [(product_name, 'latest') for product_name in product_names])
```
why the `latest` here?

nirnir93 replied:
In vscode extensions, I don't see a use case to requiring a specific version

nirolevy replied:
Because the file does not save required version - it is NOT a DB. When DB is implemented, this may change

PolaDinitz commented on src/Product.py:
```diff
@@ -0,0 +1,26 @@
+from .providers import Provider
+
+
```
do we want one or two lines between imports and class?

nirolevy replied:
2 is pep8 standard for some reason

PolaDinitz replied:
so you have some classes you didn't follow it.. that's why I asked

PolaDinitz replied:
@nirnir93

nirolevy replied:
Done

PolaDinitz commented on src/downloader.py:
```diff
@@ -1,15 +1,33 @@
-        return output_path
+    def __get_output_file_path(self, provider: Provider, product: Product) -> str:
+        package_dir = os.path.join(
+            self.__base_dir, provider.name, product.name)
```
what's with the weird /n?

PolaDinitz replied:
@nirnir93

nirolevy replied:
vscode's auto format does this somewhy. @syncush any ideas?

syncush replied:
the reason is probably line length, don't know how to fix it. I don't think it is critical in order to move on with our plans.

PolaDinitz commented on src/downloader.py:
```diff
@@ -1,15 +1,33 @@
+        package_dir = os.path.join(
+            self.__base_dir, provider.name, product.name)
+        os.makedirs(package_dir, exist_ok=True)
+        return os.path.join(package_dir, product.version + '.' + provider.file_ext)
```
`'(%d).(%d)' % product.version, provider.file_ext`

PolaDinitz replied:
@nirnir93

nirolevy replied:
I have been wondering, between this and `'{0}.{1}'.format(product.version, provider.file_ext)` which @syncush used

PolaDinitz replied:
this option will do

PolaDinitz commented on src/downloader.py:
```diff
@@ -1,15 +1,33 @@
+
+            return output_path
+        else:
+            self.__logger.log('Skipping download for product {0}@{1} from provider {2}: file already exists'.format(
```
What happens in case that the output_path isn't a file? you need to handle that case or at least the logger can't state that `file already exists`..

syncush commented on src/providers/Provider.py:
```diff
@@ -34,22 +35,39 @@ def _get_dependencies(self, product: Product, *, get_dependencies: bool = True,
+        
+        # Wait for threads to finish and kill them
+        product_queue.join()
+        for _ in threads:
+            product_queue.put(None)
```
It is unclear for me why you put `None`, is it because of line 59(in order to stop the iterator)? if so I would put it inside a function with an informative name

nirolevy replied:
It is. I'll put it in a function (or create a class for queue based thread pool?)

syncush replied:
A class with the proper doc is better imo

nirolevy replied:
Fixed

nirolevy replied:
@syncush ?

PolaDinitz commented on src/loggers/ConsoleLogger.py:
```diff
@@ -0,0 +1,5 @@
+
+class ConsoleLogger(Logger):
+    def log(self, message):
+        print(message)
```
newline

PolaDinitz commented on src/http_util.py:
```diff
@@ -0,0 +1,12 @@
+        message = 'Unknown HTTP error occurred {0}:{1}'
+
+    if message:
+        raise Exception(message.format(product_name, product_version))
```
who catches the error?

nirolevy replied:
IDK, for now no one and it crashes the cli. In a possible future where there's a REST wrapper, it should catch all errors (and then maybe errors will be different)

PolaDinitz commented on src/http_util.py:
```diff
@@ -0,0 +1,12 @@
+    message = None
+    if status_code == 404:
+        message = 'Product not found {0}:{1}'
+    if status_code > 399:
```
else if. otherwise you'll have the first message.

nirolevy replied:
Fixed

nirolevy replied:
@PolaDinitz ?

PolaDinitz commented on src/providers/Provider.py:
```diff
@@ -1,12 +1,55 @@
+from src.loggers import Logger
+
+
+def flatten(list_of_lists: List[List[object]]) -> List[object]:
```
use perhaps NumPy flatten?

syncush replied:
I would not recommend having `NumPy` as a dependency of this project, if `flatten` is the only functionality we need from it, it would be killing a bird with an F-35

nirolevy replied:
Despite the fact that a bird will never survive an encounter with an F-35, this discussion is irrelevant as this was changed and flatten is no longer used

PolaDinitz commented on src/providers/npm.py:
```diff
@@ -5,75 +5,52 @@
+        download_url = response_payload['versions'][version]['dist']['tarball']
+        return Product(self, product_name, version, download_url)
+
+    def _get_dependencies(self, product: Product, *, get_dependencies=True, get_dev_dependencies=False) -> List[Tuple[str, str]]:
```
why `get_dev_dependencies` default is false?

nirolevy replied:
Because you rarely develop a package (as opposed to using it). It also defaults to false in past and current versions of fishman

nirolevy commented:
@syncush codacy says there's a problem that I don't know how to fix... Could you help?

syncush commented:
> @syncush codacy says there's a problem that I don't know how to fix... Could you help?

I have disabled PEP 257 docs checker, I have seen a lot of threads suggesting doing it.

we should lint our code, not our docs

syncush commented:
@nirnir93, for now, add `"""This module does blah blah."""` to the top of the file.

nirolevy committed:
refactor: refactored all code

Modified cmd flags to be flags and not inputs
Moved multi process to multi thread with task queue

closes #9
closes #3
